druntime/src/core/internal/array/capacity.d

```diff
@@ -9,62 +9,150 @@
  */
 
-/// Implementation of `_d_arraysetlengthT` and `_d_arraysetlengthTTrace`
-template _d_arraysetlengthTImpl(Tarr : T[], T)
+/// Complete templated implementation of `_d_arraysetlengthT` and its GC profiling variant `_d_arraysetlengthTTrace`
+
+size_t _d_arraysetlengthT(Tarr : T[], T)(
+    return scope ref Tarr arr,
+    size_t newlength,
+    string file = __FILE__,
+    int line = __LINE__,
+    string func = __FUNCTION__
+) @trusted
 {
-    private enum errorMessage = "Cannot resize arrays if compiling without support for runtime type information!";
+    import core.lifetime : emplace;
+    import core.internal.array.utils : __arrayAlloc;
+    import core.stdc.string : memcpy, memmove;
+    import core.internal.traits : Unqual;
 
-    /**
-     * Resize dynamic array
-     * Params:
-     *  arr = the array that will be resized, taken as a reference
-     *  newlength = new length of array
-     * Returns:
-     *  The new length of the array
-     * Bugs:
-     *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
-     */
-    size_t _d_arraysetlengthT(return scope ref Tarr arr, size_t newlength) @trusted pure nothrow
+    alias U = Unqual!T; // Ensure non-inout type
+
+    if (newlength == 0)
+    {
+        arr = Tarr.init;
+        return 0;
+    }
+
+    static if (is(T == immutable) || is(T == const))
     {
-        version (DigitalMars) pragma(inline, false);
-        version (D_TypeInfo)
+        // Shrink case
+        if (newlength <= arr.length)
         {
-            auto ti = typeid(Tarr);
+            arr = arr[0 .. newlength];
+            return arr.length;
+        }
 
-            static if (__traits(isZeroInit, T))
-                ._d_arraysetlengthT(ti, newlength, cast(void[]*)&arr);
-            else
-                ._d_arraysetlengthiT(ti, newlength, cast(void[]*)&arr);
+        // Allocate new array for immutable/const
+        auto tempArr = new U[newlength];
 
-            return arr.length;
+        // Copy existing elements
+        if (arr.ptr !is null)
+        {
+            memmove(tempArr.ptr, arr.ptr, arr.length * U.sizeof);
+        }
+
+        // Ensure initialization of new elements for immutable types
+        foreach (i; arr.length .. newlength)
+        {
+            emplace(&tempArr[i], U.init);
+        }
+
+        arr = cast(Tarr) tempArr;
+        return arr.length;
+    }
+
+    // Mutable array case
+    if (newlength <= arr.length)
+    {
+        arr = arr[0 .. newlength];
+        return arr.length;
+    }
+
+    // Expand mutable array
+    size_t sizeelem = U.sizeof;
+    size_t newsize;
+    bool overflow = false;
+
+    static size_t mulu(size_t a, size_t b, ref bool overflow)
+    {
+        size_t result = a * b;
+        overflow = (b != 0 && result / b != a);
+        return result;
+    }
+
+    newsize = mulu(sizeelem, newlength, overflow);
+    if (overflow)
+    {
+        return 0;
+    }
```

Comment:
The whole point of the inline assembly code in the original hook was to verify the overflow using `setc` as opposed to that division. I'll guess the compiler can optimise your code to the same assembly, do these experiments once everything else passes.